Assets/Scripts/Models/Utility.cs

```diff
@@ -0,0 +1,784 @@
+                    {
+                        Text = "Prioritize " + Name,
+                        RequireCharacterSelected = true,
+                        Action = (ca, c) =>
+                        {
+                            c.PrioritizeJob(jobs[0]);
+                        }
+                    };
+                }
+            }
+        }
+
+        foreach (ContextMenuLuaAction contextMenuLuaAction in contextMenuLuaActions)
+        {
+            if (!contextMenuLuaAction.DevModeOnly ||
+                Settings.GetSetting("DialogBoxSettings_developerModeToggle", false))
+            {
+                yield return new ContextMenuAction
+                {
+                    Text = contextMenuLuaAction.Text,
+                    RequireCharacterSelected = contextMenuLuaAction.RequiereCharacterSelected,
+                    Action = (cma, c) => InvokeContextMenuLuaAction(contextMenuLuaAction.LuaFunction, c)
+                };
+            }
+        }
+    }
+
+    // Make a copy of the current utility.  Sub-classes should
+    // override this Clone() if a different (sub-classed) copy
+    // constructor should be run.
+    public Utility Clone()
+    {
+        return new Utility(this);
+    }
+
+    // FIXME: These functions should never be called directly,
+    // so they probably shouldn't be public functions of Utility
+    // This will be replaced by validation checks fed to use from
+    // LUA files that will be customizable for each piece of utility.
+    // For example, a door might specific that it needs two walls to
+    // connect to.
+    private bool DefaultIsValidPosition(Tile tile)
+    {
+        bool tooCloseToEdge = tile.X < MinEdgeDistance || tile.Y < MinEdgeDistance ||
+                              World.Current.Width - tile.X <= MinEdgeDistance ||
+                              World.Current.Height - tile.Y <= MinEdgeDistance;
+
+        if (tooCloseToEdge)
+        {
+            return false;
+        }
+
+        if (HasTypeTag("OutdoorOnly"))
+        {
+            if (tile.Room == null || !tile.Room.IsOutsideRoom())
+            {
+                return false;
+            }
+        }
+
+        // Make sure tile is FLOOR
+        if (tile.Type != TileType.Floor && tileTypeBuildPermissions.Contains(tile.Type) == false)
+        {
+            return false;
+        }
+
+        return true;
+    }
+
+    private void RemoveJob(Job j)
+    {
+        j.OnJobStopped -= OnJobStopped;
+        jobs.Remove(j);
+        j.buildable = null;
+    }
+
+    private void ClearJobs()
+    {
+        Job[] jobsArray = jobs.ToArray();
+        foreach (Job j in jobsArray)
+        {
+            RemoveJob(j);
+        }
+    }
+
+    private void InvokeContextMenuLuaAction(string luaFunction, Character character)
+    {
+        FunctionsManager.Utility.Call(luaFunction, this, character);
+    }
+
+    [MoonSharpVisible(true)]
+    private void UpdateOnChanged(Utility util)
```

Comment:
No idea why we make this method private if we need then need to overwrite it for Lua, and it could be used in other C# methods.